### ISSUE TITLE
Make TabsFeature return a list of TabSessionState instead a different mapped data class

### ISSUE THREAD

**Issue by jonalmeida** [mozilla-mobile/android-components#11012]

**Comment by pocmo** [comment]
Currently we also use `Tabs` to collect all the values we want to observe to only update the tabs if one of those values change:
https://github.com/mozilla-mobile/android-components/blob/main/components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt#L45-L46

**Comment by jonalmeida** [comment]
Hmm.. so you're saying if we make it `Tabs(val list: TabSessionState, val selectedTabId: String)` then we would get more updates sent to the UI that are unnecessary? 

We could mitigate this at the UI level by updating out `DiffUtils.Callback` to check the attributes that we display today. That would give us visibilty to have all the data we want from the tabs and granular control of how often it hits the main thread.

**Comment by jonalmeida** [comment]
Making `Tabs` take a list of `TabSessionState` requires `concept-tabstray` to depend on `browser-state` which isn't great..

**Comment by pocmo** [comment]
> Hmm.. so you're saying if we make it `Tabs(val list: TabSessionState, val selectedTabId: String)` then we would get more updates sent to the UI that are unnecessary?

Correct! We can work around that, by rewriting the observer. For example you could still have an `internal`-only data class that is used for collecting the values we want to observe (for `ifChanged { .. }`) and then construct whatever object you want to pass along.

(Or alternatively extend `Tab` to the properties you need, if you want to observe them too?)

> We could mitigate this at the UI level by updating out `DiffUtils.Callback` to check the attributes that we display today. That would give us visibilty to have all the data we want from the tabs and granular control of how often it hits the main thread.

I would recommend to do it in the observer, as early as possible. There's a lot of stuff in the tabs (e.g. loading percentage) and invoking the diff algorithm for all those changes would be quite wasteful.

**Comment by jonalmeida** [comment]
> Correct! We can work around that, by rewriting the observer. For example you could still have an `internal`-only data class that is used for collecting the values we want to observe (for `ifChanged { .. }`) and then construct whatever object you want to pass along.

This sounds like a great solution to our problem! We get the filtered results reduced to when only a few key items that we care about have changed, and we can return the full tab state to be consumed. If we need to alter the internal `Tab` to be triggered differently, we can still do that in AC.

> (Or alternatively extend `Tab` to the properties you need, if you want to observe them too?)

Sorry, I should have explained my use case better - we've extended `Tab` a few times where needed so far and as we've iterated within Fenix, it made sense for us to have access to the full tab state so we can display complex UI and also to re-use our [extension functions](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/tabstray/ext/TabSessionState.kt) for `TabSessionState`. With some tab functionality showing up on the homescreen tabs, it would be helpful to have functions that can act on the same type without duplication or added tranformations. For example, `fun TabSessionState.doSomething()` and `Tab.doSomething()` are the same but require some level of transformation first or querying from BrowserStore before we can apply the same function.

Another reason is to start work on converting the tabs tray to Compose, and we can start to do a bit more easily if we can get access to the tab state and follow [this meta list](https://github.com/mozilla-mobile/fenix/issues/21318) to incrementally convert it. We can convert individual tabs today and then eventually observe the store directly for drawing the whole list!
 
> I would recommend to do it in the observer, as early as possible. There's a lot of stuff in the tabs (e.g. loading percentage) and invoking the diff algorithm for all those changes would be quite wasteful.

Good point! I forget about how much "traffic" can go through a single tab.
